### ISSUE TITLE
minimap2 crashing at polishing stage

### ISSUE THREAD

**Issue by arcolon14** [fenderglass/ABruijn#152]
Greetings, 

I am running `flye v2.5-release` on ~120X PacBio reads on a ~1.1Gb fish assembly. Given our high coverage I limited the initial assembly to 50X coverage (`--asm-coverage 50`). The assembly runs without any noticeable issues until the last polishing stage (log attached bellow). I restarted the assembly and it failed with the same error. From what I understand, the `minimap2` command is failing, but I don't have any additional errors beyond:
```
[2019-09-12 00:42:25] root: ERROR: Command '['flye-minimap2', '/projects/catchenlab/angelgr2/cesox_assembly/20190909.flye2.5.asm50.oregon/40-polishing/chunks_1.fasta', '/projects/catchenlab/data/notothenioids/c.esox/pacbio/raw/m64047_190718_123258.fq.gz', '/projects/catchenlab/data/notothenioids/c.esox/pacbio/raw/m64047_190813_215709.fq.gz', '-x', 'map-pb', '-t', '52', '-a', '-p', '0.7', '-N', '10']' returned non-zero exit status -11
```
Is this just a regular `minimap2` command? If so, I was thinking on running that separately as a test. Not sure if this is a good approach or if the issue is with `flye` internally.

Just to test the assembly, I calculated statistics on the `contigs.fasta` from the `30-contigger` stage and the results appear to be pretty good. Of course, I would like to obtain the assembly from the complete `flye` run, but just in case I might keep these contigs for some data assessment. Is this last stage just polishing? I see in the documentation that it corrects some bubbles in the graphs and filters contigs of low coverage. Beyond filtering, would this stage change the assembly significantly? Also, would it be the same to polish these contigs with `arrow`/`rancon`/`pilon`/etc. Would running `flye --polish-target` be the equivalent of this last stage?

Thanks for the assistance,
Angel

Attached log file:
[flye.log](https://github.com/fenderglass/Flye/files/3606782/flye.log)

**Comment by mikolmogorov** [comment]
Hi Angel,

>Is this just a regular minimap2 command?

Yes, running the minimap2 command is a good idea - maybe the error message will be more informative. It would also help, if you could run the same command but with the latest version of minimap2 from bioconda. This way we can see if the potential error is fixed in the recent versions.

>Is this last stage just polishing?

Almost, there is polishing and scaffolding (typically, only a few joins, so does not affect the contiguity significantly). So yes, you can try to use other polishers on `contigs.fasta` too. `--polish-target` action will be almost to the remaining Flye actions, omitting scaffolding and filtering low coverage contigs (as you said).

Best,
Mikhail

**Comment by mikolmogorov** [comment]
Did you have a chance to try the minimap2 command by itself? I recently get minimap2 to crash as well, but it seems sporadic and hard to reproduce. Looks like it might depend on the overall system load.

**Comment by arcolon14** [comment]
Sorry for the (long) delay. Yes, I was able to run the `minimap2` command on its own. We didn't pursue this extensively since we found a potential solution related to coverage (see below).

Based on some of your other comments (like [this](https://github.com/lh3/minimap2/issues/489) one), I see that repetitiveness in the genome might be related to the issue. We expect our genome to be very repetitive (~50%), so it could be the exact problem.

This might not matter at all, but one thing that seemed to work for us was downsampling the data. We had very high coverage as I mentioned before. Thinking it was a memory issue, we downsampled the data to 70-80X, removing reads below a certain size. I don't think the polishing stage ever failed on that smaller dataset. We are currently working with these assemblies without any issues so far.

Thanks,
Angel

**Comment by mikolmogorov** [comment]
Thanks for the update!

Glad that it worked out in the end. It does seem like there is a bug in minimap2 - I was able to find a dataset on which it consistently crashes. Hopefully it will be fixed in the near future.

Best,
Mikhail

**Comment by dcopetti** [comment]
Hi,

I just had the same issue:
```
[2020-07-11 02:39:10] INFO: >>>STAGE: consensus
[2020-07-11 02:42:04] INFO: Running Minimap2
[2020-07-12 02:03:55] ERROR: Error running minimap2, terminating. See the alignment error log  for details: /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/minimap.stderr
[2020-07-12 02:03:55] ERROR: Command '['/bin/bash', '-c', 'set -o pipefail; flye-minimap2 /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/chunks.fasta Rabiosa_G303_35x_2kb_q7.fq.gz -x map-ont -t 94 -a -p 0.5 -N 10 --sam-hit-only -L -Q --secondary-seq | flye-samtools view -T /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/chunks.fasta -u - | flye-samtools sort -T /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/sort_200711_024204 -O bam -@ 4 -l 1 -m 4G']' returned non-zero exit status 1.
[2020-07-12 02:03:55] ERROR: Pipeline aborted
```
with a large and highly repetitive plant genome, less than 40x coverage. As far as I see in the issues, is it just a matter of restarting and hoping that the next alignment is not one of the "sometimes" when it crashes?
I get the sam output whether I run the minimap command with or without `-N 10` the sam output. Is that enough or do I have to wait for the alignment to finish?
I am running `2.7.1-b1590`. 
Thanks,
Dario

**Comment by mikolmogorov** [comment]
@dcopetti yes, if you have resources to test running minimap2 with and without `-N 10` - it will be highly appreciated. You'd need to wait until the alignment finishes, as the crash might happen somewhere in the middle.

Please use the same command as output in the log file (you might need to update paths to flye-minimap2 and flye-samtools): 
`flye-minimap2 /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/chunks.fasta Rabiosa_G303_35x_2kb_q7.fq.gz -x map-ont -t 94 -a -p 0.5 -N 10 --sam-hit-only -L -Q --secondary-seq | flye-samtools view -T /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/chunks.fasta -u - | flye-samtools sort -T /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/sort_200711_024204 -O bam -@ 4 -l 1 -m 4G`

**Comment by dcopetti** [comment]
OK, I'll wait until the assembly is at the graph writing stage :-) (it still takes >1 week in my case...) 

edit 1: it just crashed again, one day in into the alignment/bam sorting step. I will run the tests now.

edit 2: in my opinion, after reading https://github.com/lh3/minimap2/issues/489, I am not sure it will make sense to include more secondary alignments. From what I understand, SAs happen if there are two highly similar regions in the genome and both are assembled. So at the single nt level, a SA would bring in paralogous variation (if there nt variation, of course: if not it will help confirming that base) that may deviate the consensus. Basically, including SAs could be advantageous only in cases of low coverage, where the reads that actually come from that region (that will have primary alignments) are not enough (as a number) to derive a valid consensus.
Also, if some of the repeat copies were not assembled in the reference (happens with the most abundant repeats), there will be a stack of both primary and secondary alignments on top of the assembled repeat copies. Also here, I would not consider reads having a SA to compute the consensus. just my two cents, of course.
Does it make sense?

**Comment by dcopetti** [comment]
looking at the minimap.stderr file, I see that it ends with this:
```
[M::main] CMD: flye-minimap2 -x map-ont -t 94 -a -p 0.5 -N 10 --sam-hit-only -L -Q --secondary-seq /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/chunks.fasta Rabiosa_G303_35x_2kb_q7.fq.gz
[M::main] Real time: 84028.848 sec; CPU: 5923162.834 sec; Peak RSS: 87.415 GB
[bam_sort_core] merging from 1068 files and 4 in-memory blocks...
[E::hts_open_format] Failed to open file /scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/sort_200713_104107.1021.bam
samtools sort: fail to open "/scratch/dario/flye_runs/OF6p1_Rabiosa_35x_200709/10-consensus/sort_200713_104107.1021.bam": Too many open files
$ ulimit -n
1024
```
when sorting, the single bams were about 1.1 GB in size, if I remember correctly.

I asked the sys admin to increase the number of files I can write, I could also increase -m. For now I let the commands go just with and without the -N option in the minimap2 part.

**Comment by mikolmogorov** [comment]
@dcopetti you're right, I think in this case the issue is hitting the limit of opened files, not the minimap2 crash. If you can increase the number of opened files - this would resolve the issue. You can try to run `ulimit -n 4096`, but depending on your machine configuration, you might need admin rights for this. Let me know if you are able to change this, otherwise I can think about how to change alignment parameters to fix this.

Since it's a different issue, no need to test with/without -N.

**Comment by dcopetti** [comment]
@fenderglass the sys admin are not so happy of users changing the file number limit, but they let me do it for now. In my case, it may work with about 2000 files I estimate.
As of now, the files were about 1 GB in my case and probably we could afford a bit bigger files, if this cuts in number. Would changing the `-m 4G` value in the sort step help?
Can I do that (recompiling? but I installed it from conda)? or...
Thanks

**Comment by mikolmogorov** [comment]
@dcopetti yes, increasing `-m` option makes sense - this should reduce the number of temporary files. You don't need to recompile Flye - but make sure you are editing Python code that is being executed (e.g. in case you are running Flye version installed into you system). Increasing `-m` will also increase RAM consumption though - so make sure you have enough memory on the machine.

**Comment by mikolmogorov** [comment]
Hopefully this should now be finally fixed with the minimap 2.18 update in Flye 2.9

**Comment by Ciznetwork** [comment]
Issues with ruing minimap2.  I am having issues with ruing minimap2. please see the error in teh alignment log
/bin/bash: -c: line 0: unexpected EOF while looking for matching `''
/bin/bash: -c: line 1: syntax error: unexpected end of file

Someone should please help. I have also attached the complete log


Thanks

[flye.log](https://github.com/user-attachments/files/17414853/flye.log)
